design/httpproxy-status-conditions.md

```diff
@@ -0,0 +1,305 @@
+}
+
+type DetailedCondition struct {
+  Condition
+  // Errors contains a slice of relevant warning conditions for
+  // this object.
+  // Conditions are expected to appear when relevant (when there is a warning), and disappear when not relevant.
+  Errors []SubCondition `json:errors`
+  // Warnings behaves the same as Errors.
+  Warnings []SubCondition `json:warnings`
+}
+
+```
+
+The HTTPProxy `Status` struct will contain a `[]DetailedCondition` under the `conditions:` stanza.
+
+The `Valid` condition is a positive-polarity summary condition like `Ready` is on other objects, while `warnings` and `errors` are slices of abnormal-true polarity conditions that further describe problems with the configuration.
+
+The `Valid` condition must always be present when a HTTPProxy status is updated, although it may be set to either `status` `true` or `false`, meaning valid or invalid respectively.
+
+So, for a `Valid` condition with `status: true`, `errors:` must be empty, and `warnings` may have entries.
+
+If `errors` and `warnings` are empty, and `status` is true, then everything is good.
+
+`status` should not be `Unknown` for the `Valid` condition, as the updates should be effectively atomic at the end of a DAG build.
+
+If `status` is `false`, then there must be at least one entry under `errors`. If there is only one error, then that error's `reason` and `message` may be propagated up to `Valid`.
+
+This will allow us to add extra conditions covering the broad categories of HTTPProxies currently present, like so:
+
+```yaml
+---
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: tls-example
+  namespace: default
+spec:
+  routes:
+  - conditions:
+    - prefix: /
+    services:
+    - name: service-does-not-exist
+      port: 80
+  virtualhost:
+    fqdn: foo2.bar.com
+    tls:
+      secretName: testsecret-does-not-exist
+status:
+  conditions:
+  - type: Valid
+    status: false
+    observedGeneration: 1
+    lastTransitionTime: <recently>
+    reason: MultipleReasons
+    message: "Multiple reasons, see the errors stanza for more"
+    errors:
+    - type: ServiceError
+      status: true
+      reason: ServiceNotFound
+      message: "Service service-does-not-exist not found"
+    - type: TLSError
+      status: true
+      reason: TLSSecretNotFound
+      message: "TLS Secret testsecret-does-not-exist not found"
+```
+
+Or, to show an example we might do for warnings (not final, don't know if we would do this, but serves to illustrate how it would work):
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: warnings-example
+  namespace: default
+spec:
+  routes:
+  - conditions:
+    - prefix: /
+    services:
+    - name: service-has-endpoints
+      port: 80
+    - name: service-has-no-endpoints
+  virtualhost:
+    fqdn: foo2.bar.com
+    tls:
+      secretName: testsecret
+status:
+  conditions:
+  - type: Valid
+    status: true
```

Comment:
warnings with `status: true` can't happen right?

Reply:
The warning is the `ServiceError`, not the `Valid` condition. This one means "The config is partially, valid, with some problems"